File: features/step_definitions/registration.rb
Given(/^I am not logged in$/) do
    @current_page = page(MenuScreen).await
    if @current_page.loggedIn?
        @current_page.log_out
    end
    expect( @current_page.loggedIn? ).to eq(false)
end

Given(/^I am logged in as "([^"]*)"$/) do |emailAddress|
    MockServer.create_account()
    @current_page = RegistrationDriver.login( emailAddress, "password" )
    expect( @current_page.loggedIn? ).to eq(true)
end

Given(/^the user "([^"]*)" does not exist$/) do |emailAddress|
    MockServer.create_user_does_not_exist()
end

Then(/^I am registered on the server$/) do
   serverReq = JSON.parse( Mirage::Client.new.requests(2).body )
   expect( serverReq["email"] ).to eq(@emailAddress)
   expect( serverReq["name"] ).to eq("Test User")
   expect( serverReq["password"] ).to eq("t3stPassw0rd")
   expect( serverReq["group"] ).to eq(false)
   expect( serverReq["survey_consent"] ).to eq(false)
   expect( serverReq["share_name_consent"] ).to eq(false)
end

Then(/^I am logged in$/) do
    expect( @current_page.loggedIn? ).to eq(true)
end

When(/^I log in with "([^"]*)" and password "([^"]*)"$/) do |emailAddress, password|
    @current_page = RegistrationDriver.login( emailAddress, password )
end

When(/^I register as "([^"]*)"$/) do |emailAddress|
    @emailAddress = emailAddress
    @current_page = page(MenuScreen).await
    @current_page = @current_page.select("Log In")
    @current_page = page(LoginScreen).await
    @current_page = @current_page.select("Register")
    @current_page = page(RegisterScreen).await
    sleep 0.5
    @current_page = @current_page.register_via_email( emailAddress, "Test User", "t3stPassw0rd")
end

Given(/^The "([^"]*)" account exists with password "([^"]*)"$/) do |emailAddress, password|
    @emailAddress = emailAddress
    @password = password
    MockServer.create_account()
end
  
When(/^I log in via username and password$/) do
   @current_page = RegistrationDriver.login( @emailAddress, @password )
end

When(/^I restart the application$/) do
    pending # Write code here that turns the phrase above into concrete actions
  end
  
  Given(/^the server is returning a (\d+) error$/) do |arg1|
    pending # Write code here that turns the phrase above into concrete actions
  end
  
  Then(/^I get an error message$/) do
    pending # Write code here that turns the phrase above into concrete actions
  end
  
  Given(/^the user "([^"]*)" is already registered$/) do |arg1|
    pending # Write code here that turns the phrase above into concrete actions
  end
  
  
  

File: features/step_definitions/registration.js
'use strict';
const { Given, When, Then } = require('@cucumber/cucumber');
Given('I am not logged in', function() {
});

// Password is fixed to match the mock CERDI server's /token/create stub.
// `mobile: deepLink` (not driver.url, which Android routes via Chrome and
// drops custom schemes) sends the walta://login intent straight to the app.
Given('I am logged in as {string}', {timeout: 120000}, async function(emailAddress) {
    // Wait for boot first: a deeplink fired before boot is treated as a
    // cold-launch intent without the cerdiServerUrl/Secret extras, so CerdiApi
    // builds against prod and login silently never reaches the mock.
    await this.menu.waitFor();
    const url = `walta://login?email=${encodeURIComponent(emailAddress)}&password=password`;
    if (this.platform === 'android') {
        await this.driver.execute('mobile: deepLink', {
            url,
            package: 'net.thewaterbug.waterbug',
            waitForLaunch: false,
        });
    } else {
        await this.driver.execute('mobile: deepLink', {
            url,
            bundleId: 'net.thewaterbug.waterbug',
        });
    }
    await this.menu.waitForLabel("You are Logged in");
    // iOS re-opens the Menu on LOGGEDIN; let it settle to avoid a stale click.
    if (this.platform === 'ios') {
        await new Promise(r => setTimeout(r, 1500));
    }
});

Given('the user "([^"]*)" does not exist', function(emailAddress) {
});

Then('I am registered on the server', function() {
});

Then('I am logged in', {timeout: 60000}, async function() {
    await this.menu.waitForLabel("You are Logged in");
});

// 5-min timeout: form-fill + dismissSavePasswordSheet's own 60s sheet wait
// exceed the 60s budget on cold-start CI runners (warm local sims are fine).
When('I log in with {string} and password {string}', {timeout: 300000}, async function(emailAddress, password) {
    await this.menu.waitFor();
    await this.menu.login(emailAddress, password);
});

When('I register as "([^"]*)"', function(emailAddress) {
});

Given('The {string} account exists with password {string}', function(emailAddress, password) {
    global.mockCerdiServer.registerAccount({ email: emailAddress, password });
});

When('I restart the application', function() {
    return 'pending';
});

Given('the server is returning a (\d+) error', function(arg1) {
    return 'pending';
});

Then('I get an error message', function() {
    return 'pending';
});

Given('the user "([^"]*)" is already registered', function(arg1) { 
    return 'pending';
});